WordPress/gutenberg
# [Mobile] - Move Undo/Redo buttons

**Related PRs:**
- https://github.com/wordpress-mobile/gutenberg-mobile/pull/5889
- https://github.com/wordpress-mobile/WordPress-iOS/pull/20930
- https://github.com/wordpress-mobile/WordPress-Android/pull/18705

## What?
This PR updates the current location of the Undo/Redo buttons moving them from the toolbar into the app navigation header.

## Why?
It's part of the current project of updating the mobile editor's UX.

## How?
First, it removes the current location for the undo/redo buttons within the header toolbar. It now adds subscription callbacks using the React Native Bridge package to communicate with the host apps.

Updates integration undo/redo tests (editor history) by mocking the `subscribeOnUndoPressed` and `subscribeOnRedoPressed` callbacks.

## React Native Bridge

It adds two new listeners:

- `onUndoPressed` which will notify when the **undo** button was pressed from the host app's navigation header.
- `onRedoPressed` which will notify when the **redo** button was pressed from the host app's navigation header.

As well as two new functions:

- `toggleUndoButton` which will notify the host app if the **undo** button should be enabled and not dimmed.
- `toggleRedoButton` which will notify the host app if the **redo** button should be enabled and not dimmed.

These are small changes but it involves several files for both platforms.

## Device Tests Utils

It updates the `toggleHtmlMode` util to use the "More options button" (three dots) in the new header navigation bar for Android and in the already existing one for iOS.

## Demo app changes (Android)

- Customizes the MainActivity to add a navigation toolbar to the React app's view.
- Implements the Undo/Redo functionality
- Adds custom vector icons (Undo/Redo/More options)

## Demo app changes (iOS) 

- Implements the Undo/Redo functionality
- Adds custom icons (Undo/Redo/More options)
- Adds Colors asset file

## Testing Instructions

Testing should be done in **both** the Android an iOS demo apps.

- Build the demo app locally
- Open the app
- **Expect** to see the Undo/Redo buttons in the new location (Navigation header) as well as the new three dots menu.
- **Expect** the Undo/Redo buttons to be disabled and dimmed if no content change has been made
- Go over the [undo/redo test cases](https://github.com/wordpress-mobile/test-cases/tree/trunk/test-cases/gutenberg/writing-flow#undo--redo---test-cases)

## Screenshots or screencast 

<details>
    <summary>iPhone Portrait</summary>
    
Light mode | Dark mode
-|-
<kbd><img src="https://github.com/WordPress/gutenberg/assets/4885740/539e77dc-ea38-412b-8116-45a90812166b" width="250" /></kbd> | <kbd><img src="https://github.com/WordPress/gutenberg/assets/4885740/285116aa-e290-4b70-871a-223c31a81dde" width="250" /></kbd>
</details>

<details>
    <summary>iPhone Landscape</summary>
    
Light mode | Dark mode
-|-
<kbd><img src="https://github.com/WordPress/gutenberg/assets/4885740/98da8455-f5d0-47cd-9618-99499a67a735" width="250" /></kbd> | <kbd><img src="https://github.com/WordPress/gutenberg/assets/4885740/315ed062-55fa-4aa4-aee9-169233daa8d6" width="250" /></kbd>
</details>

<details>
    <summary>iPad Portrait</summary>
    
Light mode | Dark mode
-|-
<kbd><img src="https://github.com/WordPress/gutenberg/assets/4885740/465c75d1-7a72-4339-88ef-ad10c1fc8500" width="250" /></kbd> | <kbd><img src="https://github.com/WordPress/gutenberg/assets/4885740/886265da-fb39-437c-9f55-d358e3d0274f" width="250" /></kbd>
</details>

<details>
    <summary>iPad Landscape</summary>
    
Light mode | Dark mode
-|-
<kbd><img src="https://github.com/WordPress/gutenberg/assets/4885740/be6b8629-4510-47d3-83ba-76cd69025c7a" width="250" /></kbd> | <kbd><img src="https://github.com/WordPress/gutenberg/assets/4885740/edcb4858-c749-4af4-84eb-2be7f3d4994e" width="250" /></kbd>
</details>

<details>
    <summary>Android Portrait</summary>
    
Light mode | Dark mode
-|-
<kbd><img src="https://github.com/WordPress/gutenberg/assets/4885740/05ec0099-5f51-402f-8a06-66bb3557c033" width="250" /></kbd> | <kbd><img src="https://github.com/WordPress/gutenberg/assets/4885740/fa97ed6f-8183-47e4-8eb0-5b5cce38f95a" width="250" /></kbd>
</details>

<details>
    <summary>Android Lanscape</summary>
    
Light mode | Dark mode
-|-
<kbd><img src="https://github.com/WordPress/gutenberg/assets/4885740/a0b8715c-3606-469f-aae9-a96cb51cdf5a" width="250" /></kbd> | <kbd><img src="https://github.com/WordPress/gutenberg/assets/4885740/301d4100-aac9-4d1f-9a40-af9d08dcbf6e" width="250" /></kbd>
</details>

<details>
    <summary>Android Tablet Portrait</summary>
    
Light mode | Dark mode
-|-
<kbd><img src="https://github.com/WordPress/gutenberg/assets/4885740/9bc69fe1-8982-4ae1-b7bc-e0f8f596ef58" width="250" /></kbd> | <kbd><img src="https://github.com/WordPress/gutenberg/assets/4885740/7a2cbbaf-b0e3-418a-a462-e51a76ae1108" width="250" /></kbd>
</details>

<details>
    <summary>Android Tablet Lanscape</summary>
    
Light mode | Dark mode
-|-
<kbd><img src="https://github.com/WordPress/gutenberg/assets/4885740/ab07c74e-9ec6-4868-929a-a8e6664130c2" width="250" /></kbd> | <kbd><img src="https://github.com/WordPress/gutenberg/assets/4885740/95e717c2-6e6b-466b-861e-3f92f494c6d9" width="250" /></kbd>
</details>
